community-modules/client-side-row-model/src/clientSideRowModel/clientSideNodeManager.ts

```
@@ -118,32 +127,63 @@ export class ClientSideNodeManager {
-            update: [],
-            add: [],
+        const updateRowDataResult: ClientSideUpdateRowDataResult = {
+            rowNodeTransaction: { remove: [], update: [], add: [] },
+            rowsInserted: false,
         };
 
         const nodesToUnselect: RowNode[] = [];
 
-        this.executeRemove(rowDataTran, rowNodeTransaction, nodesToUnselect);
-        this.executeUpdate(rowDataTran, rowNodeTransaction, nodesToUnselect);
-        this.executeAdd(rowDataTran, rowNodeTransaction);
+        this.executeRemove(rowDataTran, updateRowDataResult, nodesToUnselect);
+        this.executeUpdate(rowDataTran, updateRowDataResult, nodesToUnselect);
+        this.executeAdd(rowDataTran, updateRowDataResult);
 
         this.updateSelection(nodesToUnselect, 'rowDataChanged');
 
-        if (rowNodeOrder) {
-            _sortRowNodesByOrder(this.rootNode.allLeafChildren, rowNodeOrder);
+        return updateRowDataResult;
+    }
+
+    /**
+     * Used by the immutable service, after updateRowData, after updating with a generated transaction to
+     * apply the order as specified by the the new data. We use positionInRootChildren to determine the order of the rows.
+     * @returns true if the order changed, otherwise false
+     */
+    public updateRowOrderFromRowData<TData>(rowData: TData[]): boolean {
+        const rows = this.rootNode.allLeafChildren;
+        if (!rows) {
+            return false;
         }
 
-        return rowNodeTransaction;
+        // A map to lookup the index of a row data
+        const newRowDataIndexMap = new Map<TData, number>();
+        rowData.forEach((data: TData, index) => {
+            newRowDataIndexMap.set(data, index);
+        });
+
+        let orderChanged = false;
+        let lastPosition = -1;
+        for (let rowIdx = 0, rowsLength = rows.length; rowIdx < rowsLength; ++rowIdx) {
+            const row = rows[rowIdx];
+
+            const newPosition = newRowDataIndexMap.get(row.data) ?? rowIdx;
```

I think the ?? rowIdx is redundant and we'd be better using `newRowDataIndexMap.get(row.data)!`

In scenarios like this we tend to let exceptions throw, as it helps to catch bad scenarios - and I can't see a scenario where we should ever _not_ have an index here